lib/src/components/NxCombobox/NxCombobox.tsx

```diff
@@ -0,0 +1,344 @@
+
+      inputClassName = classnames('nx-combobox__input', { 'nx-text-input--long': long,
+        'nx-text-input--short': short }),
+      className = classnames('nx-combobox', classNameProp, {
+        'nx-combobox--dropdown-showable': showDropdown
+      }),
+      dropdownClassName = classnames('nx-combobox__menu', {
+        'nx-combobox__menu--error': !!error
+      });
+
+  // There is a requirement that when there is an error querying the data, if the user navigates away from
+  // the component and then comes back to it the search should be retried automatically
+  function handleComponentFocus(evt: FocusEvent<HTMLDivElement>) {
+    // check if this is focus coming into the component from somewhere else on the page, not just moving between
+    // children of this component and not from focus coming into the browser from some other window
+    const comingFromOutsidePage = evt.relatedTarget === null,
+        comingFromChildNode = evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget);
+
+    if (!(comingFromOutsidePage || comingFromChildNode) && error) {
+      doSearch(searchText);
+    }
+    else if (error) {
+      setShowDropdown(true);
+    }
+  }
+
+  // In the blur event, we use the relatedTarget to check if the focused element is a child of the parent
+  // to show/hide dropdown, however, clicking the button in Safari does not focus it, which means the
+  // relatedTarget will be null, this is to stop the mouse down event being fired for the element that
+  // is listening to the blur event, in this case, the show/hide dropdown is handled by the click event
+  function handleMouseDown(evt: MouseEvent) {evt.preventDefault(); }
+
+  function handleComponentBlur(evt: FocusEvent<HTMLDivElement>) {
+    setElToFocusId('');
+    setFocusableBtnIndex(null);
+
+    // Check if the new focused element is a child of the parent, if not, then close the dropdown menu
+    if (!(evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget))) {
+      setShowDropdown(false);
+    }
+  }
+
+  function handleFilterChange(value: string) {
+    setElToFocusId('');
+    setFocusableBtnIndex(null);
+    onSearchTextChange(value);
+
+    if (value && value.trim() !== searchText.trim()) {
+      doSearch(value);
+      setPrevSearchText('');
+    }
+    else if (!value) {
+      setShowDropdown(false);
+    }
+  }
+
+  // helper for focusing different buttons in the dropdown menu
+  const adjustBtnFocus = (adjust: (i: number) => number) => () => {
+        const newFocusableBtnIndex = adjust(focusableBtnIndex ?? 0),
+            elToFocus = dropdownRef.current?.children[newFocusableBtnIndex] as HTMLElement | null;
+        if (elToFocus) {
+          setElToFocusId(elToFocus.id);
+          setFocusableBtnIndex(newFocusableBtnIndex);
+          if (autoComplete && typeof elToFocus.textContent === 'string') {
+            onSearchTextChange(elToFocus.textContent);
```

Comment:
It seems wrong to be retrieving this value from the DOM.  I'd think it should be retrieved from `matches`, probably by indexing into it with `newFocusableBtnIndex`.

Reply:
When autoComplete, the text in the input box should be updating when navigating through the options with arrow keys, could you let me know why we should not get the textcontent from the DOM element , thanks.

Reply:
> could you let me know why we should not get the textcontent from the DOM element , thanks.

Generally in React, as much as possible we want to see the DOM as an output-only medium and not as a place to store state.  This gets messed up quite a bit any time it comes to keyboard navigation because we have to get `Element` references and call `focus()` and things, but we should still try to stick to React best practices to the greatest extent that we can.  In this case that means not using the DOM to lookup the text associated with a particular id or index, but instead looking in the data structure that created that DOM in the first place (the `matches` array).

Reply:
Thanks for the clarification. It is updated here 8addbf1cdf06f65b9a25fc18185d3cc5f29ef467.